File: src/infrastructure/tools/search/search-files-by-name.ts
/**
 * search_files_by_name — 按文件名搜索
 *
 * 底层使用 Rust 的 ripgrep 核心库，性能远优于纯 JS 实现。通过 Tauri invoke 调用。
 *
 * 支持三种匹配模式：
 *   1. 纯文本（默认）：大小写不敏感子串匹配
 *   2. 正则（use_regex=true）：完整的正则表达式匹配
 *   3. Glob 模式（glob=true）：通配符匹配，如 "**\/*.ts"、"*.json"
 *      底层自动将 glob 转换为正则后由 Rust 引擎执行。
 */
import { invoke } from '@tauri-apps/api/core'
import { withCancelResult } from '@/utils/withCancel'
import { toolRegistry } from '@/domain/tools'
import type { ToolExecutor, ToolContext, ToolResult } from '@/domain/tools/types'
import { t } from '@/ui/i18n'
import { createSearchTask, globToRegex, resolveSearchRoot } from './common'

interface FileSearchResult {
  path: string
}

toolRegistry.register(
  {
    name: 'search_files_by_name',
    label: t('文件名搜索'),
    description:
      'Search for files by filename. Supports plain text (case-insensitive substring match), ' +
      'regex matching (use_regex=true), or glob patterns (glob=true, e.g. "**/*.ts", "*.json", "src/**/*.css"). ' +
      'Returns a list of matching file paths.',
    parameters: {
      type: 'object',
      properties: {
        path: {
          type: 'string',
          description: 'Root directory to search in (e.g. "./src").',
          default: '.',
        },
        query: {
          type: 'string',
          description:
            'Filename pattern. Plain text (case-insensitive) by default, regex if use_regex=true, ' +
            'glob pattern if glob=true (e.g. "**/*.ts", "*.json", "src/**/*.css").',
        },
        use_regex: {
          type: 'boolean',
          description: 'Whether query is a regex pattern. Default: false.',
          default: false,
        },
        glob: {
          type: 'boolean',
          description:
            'Whether query is a glob pattern (e.g. "**/*.ts", "*.json"). ' +
            'When true, overrides use_regex — converts glob to regex automatically. Default: false.',
          default: false,
        },
        max_results: {
          type: 'number',
          description: 'Maximum number of results. Default: 30.',
          default: 30,
        },
      },
      required: ['query'],
    },
  },
  (async (args: Record<string, any>, ctx: ToolContext): Promise<ToolResult> => {
    // 参数校验：query 为必填
    if (!args.query) {
      return {
        content:
          'Missing required parameter: "query". Please provide a filename pattern to search for.',
        uiData: { length: 0, items: [] },
      }
    }

    const { taskId, stop } = createSearchTask('search_files', ctx.abortSignal)
    let cancelled = false

    // 确定实际使用的查询内容和是否为正则模式
    let effectiveQuery = args.query
    let effectiveUseRegex = args.use_regex ?? false

    // glob 模式优先：将 glob 转换为正则，由 Rust 引擎执行
    // 当 glob=true 时忽略 use_regex（glob 转换后本身就是正则）
    if (args.glob) {
      effectiveQuery = globToRegex(args.query)
      effectiveUseRegex = true
    }

    const results: FileSearchResult[] = await withCancelResult(
      ctx.abortSignal,
      invoke<FileSearchResult[]>('search_files_by_name', {
        root: await resolveSearchRoot(args.path, ctx.sessionId),
        query: effectiveQuery,
        useRegex: effectiveUseRegex,
        maxResults: args.max_results ?? 30,
        taskId,
        // 遍历范围三个参数固定为「不剪枝」——保持工具原有行为（含隐藏项、不跳依赖目录），
        // 与 Rust 原生实现（native_tools/search/search_files_by_name.rs）逐字对齐。
        // 「默认跳过 node_modules 等目录」只是侧边栏搜索框的取舍，不作用于模型工具。
        includeHidden: true,
        skipDirNames: [],
        keepDirs: [],
      }),
      () => {
        cancelled = true
        stop()
        return [] as FileSearchResult[]
      },
    )

    if (cancelled) {
      throw `[Search cancelled] Search for "${args.query}" was cancelled.`
    }

    // 防御：invoke 可能返回 null/undefined，统一转为数组
    const safeResults: FileSearchResult[] = results ?? []

    if (safeResults.length === 0) {
      return {
        content: `No files matching "${args.query}" found in ${args.path ?? '.'}.`,
        uiData: {
          length: 0,
          items: [],
        },
      }
    }

    return {
      content:
        `🔍 ${safeResults.length} file(s) matching "${args.query}":\n` +
        safeResults.map((r) => `  📄 ${r.path}`).join('\n'),
      uiData: {
        length: safeResults.length,
        items: safeResults.map((r) => r.path),
      },
    }
  }) as ToolExecutor,
)

File: src-tauri/src/agent/native_tools/search/search_files_by_name.rs
//! `search_files_by_name` 工具（原生）— 按文件名搜索（纯文本 / 正则 / glob），带取消支持。

use crate::agent::native_tools::common::{arg_bool, arg_i64, arg_str, resolve_safe_path};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use serde_json::{json, Value};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use super::common::glob_to_regex;

pub(crate) async fn search_files_by_name_tool(
    ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let query = arg_str(args, "query").unwrap_or_default();
    if query.trim().is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: "Missing required parameter: \"query\". Please provide a filename pattern to search for.".to_string(),
            ui_data: Some(json!({ "length": 0, "items": [] })),
        });
    }

    let root = resolve_safe_path(&arg_str(args, "path").unwrap_or_else(|| ".".into()), "r", ctx.security)?;
    let use_regex = arg_bool(args, "use_regex").unwrap_or(false);
    let glob = arg_bool(args, "glob").unwrap_or(false);
    let max_results = arg_i64(args, "max_results").unwrap_or(30).clamp(1, 500) as usize;

    let mut effective_query = query.clone();
    let mut effective_use_regex = use_regex;
    if glob {
        effective_query = glob_to_regex(&query);
        effective_use_regex = true;
    }

    let cancel_flag = Arc::new(AtomicBool::new(false));
    let flag = cancel_flag.clone();
    let root_c = root.clone();
    let q = effective_query.clone();
    let task = tokio::task::spawn_blocking(move || {
        // include_hidden=true + 空剪枝清单 = **保持工具原有行为**：
        // 模型可以用显式路径 / glob 表达「就要搜 node_modules」，「默认剪掉依赖目录」
        // 是侧边栏搜索框的取舍（无表达手段），不能拿来削弱工具能力。
        crate::search::search_files_by_name(&root_c, &q, effective_use_regex, max_results, &flag, true, &[], &[])
    });

    let results = tokio::select! {
        _ = ctx.cancel.cancelled() => {
            cancel_flag.store(true, Ordering::SeqCst);
            return Ok(NativeToolOutcome::Error(format!("[Search cancelled] Search for \"{}\" was cancelled.", query)));
        }
        r = task => r.map_err(|e| format!("Search failed: {}", e))?,
    };

    let results: Vec<String> = results.into_iter().map(|r| r.path).collect();

    if results.is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: format!("No files matching \"{}\" found in {}.", query, arg_str(args, "path").unwrap_or_else(|| ".".into())),
            ui_data: Some(json!({ "length": 0, "items": [] })),
        });
    }

    let content = format!(
        "🔍 {} file(s) matching \"{}\":\n{}",
        results.len(),
        query,
        results.iter().map(|p| format!("  📄 {}", p)).collect::<Vec<_>>().join("\n")
    );

    Ok(NativeToolOutcome::Value {
        content,
        ui_data: Some(json!({ "length": results.len(), "items": results })),
    })
}
